From 06c51c792593879c19a8c34d15838b7341c94531 Mon Sep 17 00:00:00 2001
From: AppaRao Puli <apparao.puli@intel.com>
Date: Mon, 21 Nov 2022 19:18:15 +0000
Subject: [PATCH] IFWI full SPI flash update support

This commit adds the changes for allowing IFWI full
SPI flash update support. For the IFWI full SPI flash image
binary, There is no PFM to verify the image. So check
if image size is 64MB (BIOS SPI flash size), then assume
its BIOS image and skip the verification.

Tested:
- Control successfuly passed to fwupd.sh

Signed-off-by: AppaRao Puli <apparao.puli@intel.com>
---
 pfr_image_manager.cpp | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/pfr_image_manager.cpp b/pfr_image_manager.cpp
index 2bc6420..c04e1f9 100644
--- a/pfr_image_manager.cpp
+++ b/pfr_image_manager.cpp
@@ -22,6 +22,7 @@
 #include <cstring>
 #include <filesystem>
 #include <fstream>
+#include <iostream>
 #include <iomanip>
 #include <set>
 #include <sstream>
@@ -303,11 +304,26 @@ int Manager::processImage(const std::string& imgFilePath)
     }
 
     int retry = 3;
-    std::string ver;
     std::string extVer;
+    std::string ver;
     std::string purposeString;
 
-    if (0 != verifyImage(imgFilePath, ver, purposeString))
+    // For the IFWI full SPI flash image binary, There is no PFM to
+    // verify the image. So check if image size is 64MB (BIOS SPI
+    // flash size), then assume its BIOS image and skip verification.
+    std::uintmax_t fSize = std::filesystem::file_size(imgFilePath);
+    std::cout << "File: " << imgFilePath << ", Size: " << fSize << "\n";
+    if (fSize == 64 * 1024 * 1024)
+    {
+        // TODO: Find out how to read version number and also set correct
+        // "purpose" string. Filling with dummy data for now.
+        phosphor::logging::log<phosphor::logging::level::ERR>(
+            "IFWI full flash update request");
+        ver = "1";
+        purposeString =
+            "xyz.openbmc_project.Software.Version.VersionPurpose.Host";
+    }
+    else if (0 != verifyImage(imgFilePath, ver, purposeString))
     {
         phosphor::logging::log<phosphor::logging::level::ERR>(
             "Error verifying uploaded image");
-- 
2.17.1

